pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,507 @@
+	}
+	psm.cloudEvents = psm.cloudEvents[:psm.length]
+	psm.rawData = psm.rawData[:psm.length]
+	psm.entries = psm.entries[:psm.length]
+	envelope := runtimePubsub.NewBulkSubscribeEnvelope(&runtimePubsub.BulkSubscribeEnvelope{
+		ID:       id,
+		Topic:    topic,
+		Entries:  psm.rawData,
+		Pubsub:   psName,
+		Metadata: msg.Metadata,
+	})
+	da, marshalErr := json.Marshal(&envelope)
+	if marshalErr != nil {
+		log.Errorf("error serializing bulk cloud event in pubsub %s and topic %s: %s", psName, msg.Topic, marshalErr)
+		if route.deadLetterTopic != "" {
+			entries := make([]pubsub.BulkMessageEntry, len(psm.entries))
+			for i, entry := range psm.entries {
+				entries[i] = *entry
+			}
+			bulkMsg := pubsub.BulkMessage{
+				Entries:  entries,
+				Topic:    msg.Topic,
+				Metadata: msg.Metadata,
+			}
+			if dlqErr := a.sendBulkToDeadLetter(ctx, psName, &bulkMsg, route.deadLetterTopic, entryIDIndexMap, nil); dlqErr == nil {
+				// dlq has been configured and message is successfully sent to dlq.
+				diag.DefaultComponentMonitoring.BulkPubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), msg.Topic, 0)
+				for _, item := range psm.entries {
+					ind := (*entryIDIndexMap)[item.EntryID]
+					(*bulkResponses)[ind].EntryID = item.EntryID
+					(*bulkResponses)[ind].Error = nil
+				}
+				return nil
+			}
+		}
+		diag.DefaultComponentMonitoring.BulkPubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Retry)), msg.Topic, 0)
+		for _, item := range psm.entries {
+			ind := (*entryIDIndexMap)[item.EntryID]
+			(*bulkResponses)[ind].EntryID = item.EntryID
+			(*bulkResponses)[ind].Error = marshalErr
+		}
+		return marshalErr
+	}
+	psm.data = da
+	psm.path = path
+	return policy(func(ctx context.Context) error {
+		switch a.runtimeConfig.ApplicationProtocol {
+		case HTTPProtocol:
+			return a.publishBulkMessageHTTP(ctx, &psm, bulkResponses, *entryIDIndexMap)
+		default:
+			return backoff.Permanent(errors.New("invalid application protocol"))
+		}
+	})
+}
+
+// publishBulkMessageHTTP publishes bulk message to a subscriber using HTTP and takes care of corresponding response.
+func (a *DaprRuntime) publishBulkMessageHTTP(ctx context.Context, msg *pubsubBulkSubscribedMessage,
+	bulkResponses *[]pubsub.BulkSubscribeResponseEntry, entryIDIndexMap map[string]int,
+) error {
+	spans := make([]trace.Span, len(msg.entries))
+
+	req := invokev1.NewInvokeMethodRequest(msg.path)
+	req.WithHTTPExtension(nethttp.MethodPost, "")
+	req.WithRawData(msg.data, contenttype.CloudEventContentType)
+	req.WithCustomHTTPMetadata(msg.metadata)
+
+	n := 0
+	for _, cloudEvent := range msg.cloudEvents {
+		if cloudEvent[pubsub.TraceIDField] != nil {
+			traceID := cloudEvent[pubsub.TraceIDField].(string)
+			sc, _ := diag.SpanContextFromW3CString(traceID)
+			spanName := fmt.Sprintf("pubsub/%s", msg.topic)
+			var span trace.Span
+			ctx, span = diag.StartInternalCallbackSpan(ctx, spanName, sc, a.globalConfig.Spec.TracingSpec)
```

Comment:
I found this one working well in case of normal subscription and for tracing, my use case is almost similar, where-in I trace against the published traceID only, if any. I found this working well in bulk subscription case as well.